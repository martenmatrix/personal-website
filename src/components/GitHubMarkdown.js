import ReactMarkdown from 'react-markdown';
import makeEmojisAccessible from '@fec/remark-a11y-emoji'
import parseGitHubEmojis from 'remark-gemoji';
import supportGFM from 'remark-gfm';
import addIdToHeadings from 'rehype-slug';
import sanitizeHTML from 'rehype-sanitize';
import parseHTML from 'rehype-raw';
import removeHeading from 'remark-first-heading';
import { visit } from 'unist-util-visit';
import '../styles/github-markdown-light.css';
function fixBrokenLinks() {
    // This code is from the documentation of rehype-sanitize: https://github.com/rehypejs/rehype-sanitize#example

    // Page load (you could wrap this in a DOM ready if the script is loaded early).
    hashchange()

    // When URL changes.
    window.addEventListener('hashchange', hashchange)

    // When on the URL already, perhaps after scrolling, and clicking again, which
    // doesnâ€™t emit `hashchange`.
    document.addEventListener(
      'click',
      (event) => {
        if (
          event.target &&
          event.target instanceof HTMLAnchorElement &&
          event.target.href === window.location.href &&
          window.location.hash.length > 1
        ) {
          setTimeout(() => {
            if (!event.defaultPrevented) {
              hashchange()
            }
          })
        }
      },
      false
    )

    function hashchange() {
      /** @type {string|undefined} */
      let hash

      try {
        hash = decodeURIComponent(window.location.hash.slice(1)).toLowerCase()
      } catch {
        return
      }

      const name = 'user-content-' + hash
      const target =
        document.getElementById(name) || document.getElementsByName(name)[0]

      if (target) {
        setTimeout(() => {
          target.scrollIntoView()
        }, 0)
      }
    }
}
fixBrokenLinks();

function GitHubMarkdown({ children }) {
    // Remark plugins to transform emojis is passed into rehypePlugins array => ids need to be created before github emojis are parsed, otherwise toc will not work
    return <ReactMarkdown className="markdown-body" remarkPlugins={[supportGFM, makeEmojisAccessible, [removeHeading, {heading: ''}]]} rehypePlugins={[addIdToHeadings, parseGitHubEmojis, parseHTML, sanitizeHTML]}>{children}</ReactMarkdown>
}

export default GitHubMarkdown;