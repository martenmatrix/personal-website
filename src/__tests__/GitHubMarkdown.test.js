import { render, screen } from '@testing-library/react';
import GitHubMarkdown from "../components/GitHubMarkdown";

const markdown = `
# Cocktail Mixer
This contains some GitHub-specific markdown.

## Table of Contents
- [Usage](#grey_exclamation-usage)
- [Features](#sparkles-features)

## :grey_exclamation: Usage
<p align="center">
	<img alt="homepage of frontend with numbers assigned to the sections" src="https://github.com/martenmatrix/cocktail-mixer/blob/main/readme-images/homepage.jpg?raw=true" />
</p>
<ol>
	<li>
		lorum lorum sorum
	</li>
	<li>
        lorum lorum sorum 2
	</li>
	<li>
		<details>
			<summary>
				This will open up the settings. You will be prompted for a password. The default password is <code>cocktail</code>.
			</summary>
			<div>
				<ol>
					<p align="center">
						<img alt="two images: static settings site with numbers on the left and animated settings site with action on the right" src="https://raw.githubusercontent.com/martenmatrix/cocktail-mixer/main/readme-images/static-and-animation-settings.gif" />
					</p>
					<li>
						Add an ingredient and assign it a category. The current categories consist of <code>juice</code>, <code>softdrink</code>, <code>alcohol</code> and <code>unableToPump</code>.
					</li>
				</ol>
			</div>
		</details>
	</li>
</ol>

## :sparkles: Features
- everyone on your local network can make a drink by visiting the website hosted by the machine in your local network
- automatically mix a drink with the press of a button

1. First list item.
1. Second one.

## :wrench: Installation

> :warning: The following guide is for Linux distributions.
1. Clone the repository.
2. Navigate into the folder:

#### Cool code
\`const cool = (text) => console.log(text)\`
`

test('parses markdown correctly', () => {
    const { container } = render((
    <GitHubMarkdown>
        {markdown}
    </GitHubMarkdown>))

    expect(container).toMatchSnapshot();
});