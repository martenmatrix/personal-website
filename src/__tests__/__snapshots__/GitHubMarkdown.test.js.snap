// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses markdown correctly 1`] = `
<div>
  <div
    class="markdown-body"
  >
    <h1
      id="user-content-"
    />
    

    <p>
      This contains some GitHub-specific markdown.
    </p>
    

    <h2
      id="user-content-table-of-contents"
    >
      Table of Contents
    </h2>
    

    <ul>
      

      <li>
        <a
          href="#grey_exclamation-usage"
        >
          Usage
        </a>
      </li>
      

      <li>
        <a
          href="#sparkles-features"
        >
          Features
        </a>
      </li>
      

    </ul>
    

    <h2
      id="user-content-grey_exclamation-usage"
    >
      ‚ùï Usage
    </h2>
    

    <p
      align="center"
    >
      
	
      <img
        alt="homepage of frontend with numbers assigned to the sections"
        src="https://github.com/martenmatrix/cocktail-mixer/blob/main/readme-images/homepage.jpg?raw=true"
      />
      

    </p>
    

    <ol>
      
	
      <li>
        
		lorum lorum sorum
	
      </li>
      
	
      <li>
        
        lorum lorum sorum 2
	
      </li>
      
	
      <li>
        
		
        <details>
          
			
          <summary>
            
				This will open up the settings. You will be prompted for a password. The default password is 
            <code>
              cocktail
            </code>
            .
			
          </summary>
          
			
          <div>
            
				
            <ol>
              
					
              <p
                align="center"
              >
                
						
                <img
                  alt="two images: static settings site with numbers on the left and animated settings site with action on the right"
                  src="https://raw.githubusercontent.com/martenmatrix/cocktail-mixer/main/readme-images/static-and-animation-settings.gif"
                />
                
					
              </p>
              
					
              <li>
                
						Add an ingredient and assign it a category. The current categories consist of 
                <code>
                  juice
                </code>
                , 
                <code>
                  softdrink
                </code>
                , 
                <code>
                  alcohol
                </code>
                 and 
                <code>
                  unableToPump
                </code>
                .
					
              </li>
              
				
            </ol>
            
			
          </div>
          
		
        </details>
        
	
      </li>
      

    </ol>
    

    <h2
      id="user-content-sparkles-features"
    >
      ‚ú® Features
    </h2>
    

    <ul>
      

      <li>
        everyone on your local network can make a drink by visiting the website hosted by the machine in your local network
      </li>
      

      <li>
        automatically mix a drink with the press of a button
      </li>
      

    </ul>
    

    <ol>
      

      <li>
        First list item.
      </li>
      

      <li>
        Second one.
      </li>
      

    </ol>
    

    <h2
      id="user-content-wrench-installation"
    >
      üîß Installation
    </h2>
    

    <blockquote>
      

      <p>
        ‚ö†Ô∏è The following guide is for Linux distributions.
      </p>
      

    </blockquote>
    

    <ol>
      

      <li>
        Clone the repository.
      </li>
      

      <li>
        Navigate into the folder:
      </li>
      

    </ol>
    

    <h4
      id="user-content-cool-code"
    >
      Cool code
    </h4>
    

    <p>
      <code>
        const cool = (text) =&gt; console.log(text)
      </code>
    </p>
  </div>
</div>
`;
